src/librustc/ty/layout.rs

```diff
@@ -1517,10 +1517,19 @@ impl<'a, 'tcx> LayoutCx<'tcx, TyCtxt<'a, 'tcx, 'tcx>> {
                             let offset = st[i].fields.offset(field_index) + offset;
                             let size = st[i].size;
 
-                            let abi = if offset.bytes() == 0 && niche.value.size(dl) == size {
-                                Abi::Scalar(niche.clone())
-                            } else {
-                                Abi::Aggregate { sized: true }
+                            let abi = match st[i].abi {
+                                Abi::Scalar(_) => Abi::Scalar(niche.clone()),
+                                Abi::ScalarPair(ref first, ref second) => {
+                                    if offset.bytes() == 0 {
+                                        // We need to use scalar_unit to reset the
+                                        // valid range to the maximal one for that
+                                        // primitive.
```

Comment:
Can you put it outside and say that it's because only the niche is guaranteed to always be initialized, not the other primitive?

Reply:
Amended and completed.